### PR TITLE
feat(pi05): add optional visual and proprioceptive MEM

### PR DESCRIPTION
**Pull request by pkooij**
## Summary

Adds optional short-horizon visual and proprioceptive memory to Pi05, based on MEM.

Both paths are independently opt-in through `policy.use_visual_memory` and `policy.use_proprioceptive_memory`. Their defaults are `false`, so existing Pi05 checkpoints and configurations retain single-frame behavior.

## What changed

- Fuse historical image frames inside SigLIP with causal temporal attention between matching patch tokens.
- Return only current-frame vision tokens so the downstream language/action prefix length stays unchanged.
- Represent each retained proprioceptive state with one projected continuous token.
- Add configurable frame count, stride, temporal-attention frequency, episode reset, masked history padding, and strided inference queues.
- Preserve the new proprioceptive projection when initializing from a non-MEM Pi05 checkpoint.
- Resolve image and state history timestamps independently, including through dataset rename maps.
- Preserve temporal padding metadata and renamed normalization statistics through preprocessing.
- Include the RoboMME camera/task/resource-lifecycle fixes needed for memory-enabled Pi05 training and evaluation.

## Usage

```bash
--policy.use_visual_memory=true \
--policy.use_proprioceptive_memory=true \
--policy.memory_frames=6 \
--policy.memory_stride=10
```

Enable either memory path independently. Enabling neither leaves standard Pi05 behavior unchanged.

## RoboMME results

Both policies were initialized from `lerobot/pi05_base` and trained on `lerobot/robomme` for 30,000 steps using 8× H100 GPUs, batch size 24 per GPU (global batch size 192), flow-only training, and no subtasks. The benchmark used 50 task IDs for each of RoboMME’s 16 task groups: 800 matched episodes per policy and 1,600 rollouts total.

| Policy | Successes | Success rate |
| --- | ---: | ---: |
| Pi05 baseline (no memory) | 15 / 800 | 1.88% |
| Pi05 + visual and proprioceptive MEM | 146 / 800 | 18.25% |

MEM improved success by **+16.38 percentage points** and achieved a **9.73×** relative success rate. A paired bootstrap over matched episodes gives a 95% confidence interval of **[+13.63, +19.12] percentage points**. There were 138 MEM-only successes, 7 baseline-only successes, 8 successes from both policies, and 647 failures from both policies (exact paired McNemar `p ≈ 1.09e-32`).

| RoboMME task group | Baseline | Visual + proprio MEM |
| --- | ---: | ---: |
| BinFill | 0 / 50 (0%) | 8 / 50 (16%) |
| ButtonUnmask | 0 / 50 (0%) | 12 / 50 (24%) |
| ButtonUnmaskSwap | 0 / 50 (0%) | 0 / 50 (0%) |
| InsertPeg | 0 / 50 (0%) | 0 / 50 (0%) |
| MoveCube | 6 / 50 (12%) | 22 / 50 (44%) |
| PatternLock | 5 / 50 (10%) | 6 / 50 (12%) |
| PickHighlight | 0 / 50 (0%) | 14 / 50 (28%) |
| PickXtimes | 0 / 50 (0%) | 14 / 50 (28%) |
| RouteStick | 2 / 50 (4%) | 6 / 50 (12%) |
| StopCube | 2 / 50 (4%) | 4 / 50 (8%) |
| SwingXtimes | 0 / 50 (0%) | 5 / 50 (10%) |
| VideoPlaceButton | 0 / 50 (0%) | 12 / 50 (24%) |
| VideoPlaceOrder | 0 / 50 (0%) | 13 / 50 (26%) |
| VideoRepick | 0 / 50 (0%) | 5 / 50 (10%) |
| VideoUnmask | 0 / 50 (0%) | 9 / 50 (18%) |
| VideoUnmaskSwap | 0 / 50 (0%) | 16 / 50 (32%) |

These are single-training-seed results; additional seeds are needed to estimate training variance.

## Validation

- Focused Pi05 MEM, rename-processor, and RoboMME suites: `51 passed, 2 skipped`.
- Targeted pre-commit hooks passed, including Ruff, Prettier, typos, Bandit, and mypy.

### CODE REVIEW

**Comment by pkooij** [comment]
@claude please review

File: src/lerobot/policies/pi05/memory.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+# Copyright 2026 The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (s1lent4gnt) [inline src/lerobot/policies/pi05/memory.py:7]: Please use LeRobot licence template!

File: src/lerobot/policies/pi05/memory.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        if num_frames == 1 or (layer_index + 1) % temporal_attention_every:
+            layer_output = _hidden_tensor(
+                layer(spatial_input, attention_mask=None),
+                source=f"SigLIP encoder layer {layer_index}",
+            )
+            hidden_states = layer_output.reshape(batch_size, num_frames, num_patches, hidden_size)
+            continue
+
+        residual = hidden_states
+        spatial_norm = layer.layer_norm1(spatial_input)
```
Comment (s1lent4gnt) [inline src/lerobot/policies/pi05/memory.py]: In paper appendix C adds temporal position encoding to spatial input. To be faithful to paper could you add it as well.

File: src/lerobot/policies/pi05/processor_pi05.py
Comment (s1lent4gnt) [inline src/lerobot/policies/pi05/processor_pi05.py]: In Section III-D of the paper, they use only the proprioceptive state with a linear projection. The state discretization is removed from the prompt, preventing the state from being duplicated.

File: tests/policies/pi0_pi05/test_pi05_memory.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
+# Copyright 2026 The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (s1lent4gnt) [inline tests/policies/pi0_pi05/test_pi05_memory.py:7]: Could you update license here also

**Review by s1lent4gnt — COMMENTED** [review]
Good overall PR. Before merging, I’d like to address these three points:

- Refactor rename_map: rename once at the raw-batch boundary:
Right now renaming happens in several places (timestamps, stats, train loop, observation processor). I suggest one serialized processor step that renames the raw batch before `batch_to_transition()`,

- PEFT: proprio_history_proj not covered by default PEFT config:
When `use_proprioceptive_memory=true`, `proprio_history_proj` is a new layer that does not exist in `lerobot/pi05_base`. Under PEFT, it is neither in `target_modules` nor in `modules_to_save`, so it stays frozen and is not saved in adapter checkpoints.

- `robomme.mdx` out of sync with implementation:
`robomme.mdx` still documents `observation.images.image` / `wrist_image` and says no renaming is needed. The code defaults to `camera1` / `camera2` (RoboMMEEnv, tests, eval observations). Please align docs with the code.

File: docs/source/pi05.mdx
Comment (s1lent4gnt) [inline docs/source/pi05.mdx]: Since `robomme.mdx` now correctly documents fine-tuning with both `--policy.pretrained_path` and `--rename_map`. Could update this as well?

File: src/lerobot/envs/configs.py [hunk 4]
```diff
@@ -862,24 +862,29 @@ class RoboMMEEnv(EnvConfig):
...
+            front_camera_key: PolicyFeature(type=FeatureType.VISUAL, shape=(256, 256, 3)),
+            wrist_camera_key: PolicyFeature(type=FeatureType.VISUAL, shape=(256, 256, 3)),
             "agent_pos": PolicyFeature(type=FeatureType.STATE, shape=(8,)),
         }
+        self.features_map = {
+            ACTION: ACTION,
+            front_camera_key: f"{OBS_IMAGES}.{self.front_camera_name}",
+            wrist_camera_key: f"{OBS_IMAGES}.{self.wrist_camera_name}",
+            "agent_pos": OBS_STATE,
+        }
```
Comment (s1lent4gnt) [inline src/lerobot/envs/configs.py]: This overwrite any mapping provided through the config and an explicitly supplied mapping is silently discarded.

File: src/lerobot/scripts/lerobot_train.py [hunk 5]
```diff
@@ -715,6 +717,7 @@ def train(cfg: TrainPipelineConfig):
         for cam_key in dataset.meta.camera_keys:
             if cam_key in batch and batch[cam_key].dtype == torch.uint8:
                 batch[cam_key] = batch[cam_key].to(dtype=torch.float32) / 255.0
+        batch = rename_batch_keys(batch, cfg.rename_map)
```
Comment (s1lent4gnt) [inline src/lerobot/scripts/lerobot_train.py]: Could we apply the same raw-batch renaming in the held-out eval loop before `preprocessor(eval_batch)`?

File: src/lerobot/policies/factory.py
Comment (s1lent4gnt) [inline src/lerobot/policies/factory.py]: `features` is renamed from `actions` to `action`, but `ds_meta.features` still keeps the raw `actions` key. This lookup therefore returns no action names. Could we use `rename_map` to find the raw action key before reading its `names` metadata? This is needed for exclusions such as `gripper` when relative actions are enabled.

**Comment by pkooij** [comment]
@s1lent4gnt addressed all four:
1. Updated the Pi05 docs with --rename_map with --policy.pretrained_path.
2. RoboMME’s generated feature mappings is added to defaults without overwriting.
3. Training and eval share the same  preprocessing path (and thus renaming) 
4. Action feat names are read from the raw dataset key via rename_map

File: src/lerobot/policies/pi05/memory.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        raise ValueError(
+            f"frame_mask must have shape {(batch_size, num_frames)}, got {tuple(frame_mask.shape)}"
+        )
+    if any(not hasattr(vision_model, name) for name in ("embeddings", "encoder", "post_layernorm")):
+        raise TypeError("MEM visual memory requires a SigLIP-compatible vision transformer")
+    if not hasattr(vision_model.encoder, "layers"):
+        raise TypeError("MEM visual memory requires a SigLIP encoder exposing a layers collection")
+
+    layers = vision_model.encoder.layers
+    temporal_layers = [i for i in range(len(layers)) if (i + 1) % temporal_attention_every == 0]
```
Comment (s1lent4gnt) [inline src/lerobot/policies/pi05/memory.py:216]: We should check `temporal_attention_every` to be less then SigLip number of layers, otherwise `temporal_layers` will be empty.

**Comment by pkooij** [comment]
@s1lent4gnt results now, (only 5k steps, 6x less training than 18% succesrate of 30k steps, but still higher than 2% non MEM)

75/800 successes = 9.375%.
-------------------
Permanence: 15%
Reference: 11%
Counting: 7%
Imitation: 4.5%

**Review by s1lent4gnt — APPROVED** [review]
Nice work!
We can now train a policy that can play "Three-Cup Monte"
